### ISSUE TITLE
Could not reproduce results on GSM using olmoe

### ISSUE THREAD

**Issue by neilwen987** [yaof20/DenseMixer#10]
I can't reproduce your results on GSM using olmoe model.
I run the following command, using densemixer.
```
bash scripts/train/finetune/full.sh --task gsm --model olmoe --total_batch_size 256 --num_train_epochs 4 --devices 0,1,2,3,4,5,6,7 --num_gpus 8 --port 29011 --lr 2e-5 --per_device_train_batch_size 4 --per_device_eval_batch_size 4
```

And I do evaluation on GSM, get results 44.5 which is much lower than 49.00 reported in the blog.
Any suggestions?:)

**Comment by yaof20** [comment]
Hi,

Thanks for bringing this up!

We've identified two main reasons for the discrepancy:

- `**Learning Rate:**` The released script for OLMoE on GSM used an incorrect learning rate. It should be 1e-6, not 2e-5. We're currently auditing and updating all scripts in the repo to reflect the correct settings.

- `**Code Reorganization & Framework Differences:**` In our experiments, we directly modified the original libraries -- Open-Instruct for OLMoE and LLaMAFactory for the other two models. When preparing the release, we reorganized the code into a unified PyPI package, which introduced some mismatches between the released and experimental code. This added complexity and led to several issues. We're now **actively aligning everything** to ensure accurate and reliable reproducibility.

So far, using the released code:
- `**Qwen1.5-MoE-A2.7B:**` Reproduced 54.97 on GSM (vs. 55.16 in the blog; baseline: 53.42 in the blog)
- `**OLMoE** (with corrected learning rate):` Reproduced 46.3 (vs. 49.00 in the blog; baseline: 45.94 in the blog).
(We're still investigating this gap. Please allow some extra time, as Open-Instruct trains slower and we're currently using fewer GPUs to fully replicate the original setup.)

We are **actively working on fixing this** and we will definitely get back to you soon. The repo will be updated shortly.

In the meantime, we recommend starting with the **14B** model, as its released implementation has been preliminarily verified and closely matches our reported results. The 7B version is still under **active debugging and refinement**.

Thanks again for your patience and feedback — please stay tuned!

**Comment by yaof20** [comment]
> I can't reproduce your results on GSM using olmoe model. I run the following command, using densemixer.
> 
> ```
> bash scripts/train/finetune/full.sh --task gsm --model olmoe --total_batch_size 256 --num_train_epochs 4 --devices 0,1,2,3,4,5,6,7 --num_gpus 8 --port 29011 --lr 2e-5 --per_device_train_batch_size 4 --per_device_eval_batch_size 4
> ```
> 
> And I do evaluation on GSM, get results 44.5 which is much lower than 49.00 reported in the blog. Any suggestions?:)

Hi,

Thanks for bringing this up!

We've identified two main reasons for the discrepancy:

**Learning Rate:** The released script for OLMoE on GSM used an incorrect learning rate. It should be 1e-6, not 2e-5. We're currently auditing and updating all scripts in the repo to reflect the correct settings.

**Code Reorganization & Framework Differences:** In our experiments, we directly modified the original libraries -- Open-Instruct for OLMoE and LLaMAFactory for the other two models. When preparing the release, we reorganized the code into a unified PyPI package, which introduced some mismatches between the released and experimental code. This added complexity and led to several issues. We're now actively aligning everything to ensure accurate and reliable reproducibility.

So far, using the released code:

**Qwen1.5-MoE-A2.7B:** Reproduced 54.97 on GSM (vs. 55.16 in the blog; baseline: 53.42 in the blog)
**OLMoE** (with corrected learning rate): Reproduced 46.3 (vs. 49.00 in the blog; baseline: 45.94 in the blog).
(We're still investigating this gap. Please allow some extra time, as Open-Instruct trains slower and we're currently using fewer GPUs to fully replicate the original setup.)
We are actively working on fixing this and we will definitely get back to you soon. The repo will be updated shortly.

In the meantime, we recommend starting with the 14B model, as its released implementation has been preliminarily verified and closely matches our reported results. The 7B version is still under active debugging and refinement.

Thanks again for your patience and feedback — please stay tuned!

**Comment by neilwen987** [comment]
Can you share the traning curve of olmoe using densemixer? I tried the corrected lr and found that the eval ppl keeps increasing during training. It seems a little weird🥲.

<img width="372" height="617" alt="Image" src="https://github.com/user-attachments/assets/152ed9a3-104a-4cce-b63c-8b98bfa0be3d" />

**Comment by neilwen987** [comment]
> https://wandb.ai/experiments_-university-of-california-san-diego/open_instruct_moe_gsm/reports/olmoe-gsm-run-lr-1e-6--VmlldzoxMzYzMjg5NQ?accessToken=677we11cmjiog5dx19180bnn0njrbxs38ec5wy5fgqu97ks52ktlvg6x63syaop8

Thanks for sharing the curve!  I wonder the number of GPUS may affect the training process ? 
I'm using 8 A100 with 40GB and set the aruguments as :
```
--total_batch_size 256  --per_device_train_batch_size 4
```
which means that I update the gradient every 8 steps instead of 32 steps (default setting using 2 GPUS). I tried to set the 
```
--per_device_train_batch_size 1
```
as this would allow using the gradient updates every 32 steps but it tooks to long to get results (~8 hours)😭
I found that in my setting, the final training loss is 0.285 vs 0.246 (provided) and the performance on GSM is 45.64 vs 46.3.
So I guess maybe MoE is very senstive to training environments? Any suggestions? How did you guys find the best training parameters? ：）

**Comment by yaof20** [comment]
> > https://wandb.ai/experiments_-university-of-california-san-diego/open_instruct_moe_gsm/reports/olmoe-gsm-run-lr-1e-6--VmlldzoxMzYzMjg5NQ?accessToken=677we11cmjiog5dx19180bnn0njrbxs38ec5wy5fgqu97ks52ktlvg6x63syaop8
> 
> Thanks for sharing the curve! I wonder the number of GPUS may affect the training process ? I'm using 8 A100 with 40GB and set the aruguments as :
> 
> ```
> --total_batch_size 256  --per_device_train_batch_size 4
> ```
> 
> which means that I update the gradient every 8 steps instead of 32 steps (default setting using 2 GPUS). I tried to set the
> 
> ```
> --per_device_train_batch_size 1
> ```
> 
> as this would allow using the gradient updates every 32 steps but it tooks to long to get results (~8 hours)😭 I found that in my setting, the final training loss is 0.285 vs 0.246 (provided) and the performance on GSM is 45.64 vs 46.3. So I guess maybe MoE is very senstive to training environments? Any suggestions? How did you guys find the best training parameters? ：）

Hi,

Thanks for replying. 

To clarify -- The curve I shared was actually reproduced on 8 × H200 GPUs with `gradient checkpointing` enabled. We tested all checkpoints and reported the highest one (not simply testing the final checkpoint). *(Sorry for not making it clear that we were also experimenting on 2 GPUs. And the 49.00 result in the blog came from our very early 7B-model experiments on **2xA100 GPUs**, before we scaled up to the 14B and 30B runs. We are still checking this.)*

For your question: 

I think it’s not that “MoE is sensitive to training environments,” but rather that **hyper‑parameters matter** (even for dense models)—and we **didn’t cherry‑pick** settings to favor DenseMixer. Here’s our protocol:

1. **Baseline Grid Search**  
   - Run the *full-finetune* baseline over a grid of learning rates and batch sizes.  
   - Save checkpoints at each epoch and **evaluate every checkpoint** on the test set.  
   - **Record the best** validation score as the baseline. (Also note down the corresponding training hyper-parameters.)

2. **DenseMixer Under Strictly Controlled Conditions**  
   - Use **exactly the same** training hyper-parameters and settings (e.g., the GPU number and GPU type) as the baseline.  
   - Again, **evaluate all checkpoints** and report the highest score.  

By holding everything constant (we did spend a lot of time here), we ensure any gains are due solely to DenseMixer, not experimental variance.

If your goal is simply to validate DenseMixer’s benefits, you don’t need to reproduce our blog’s exact numbers. Instead, run a **controlled comparison**—use identical settings for both the baseline and DenseMixer—and you’ll clearly see the relative improvement.

We’re currently verifying all results with our released code and will update this thread as soon as everything is confirmed. In the meantime, feel free to ask any further questions. For real‑time discussion or even to schedule a Zoom call, you can reach me on WeChat: `yaof20`.  :)

**Comment by yaof20** [comment]
Hi, 

We have rerun the 7B and 14B experiments using the running scripts provided in the current repo. And the results are as follows, which are slightly different from the blog, but still lead to the same conclusion that DenseMixer is bringing improvement.

**7B-Results:**
Method | GSM | MBPP | HumanEval | ESFT-intent | ESFT-law | ESFT-summary | ESFT-translation
-- | -- | -- | -- | -- | -- | -- | --
Conventional | 0.4594 | 0.2320 | 0.1341 | 0.7260 | 0.2500 | 0.3370 | 0.2756
DenseMixer | 0.4700 | 0.2440 | 0.1644 | 0.7540 | 0.2790 | 0.3750 | 0.3032

**14B-Results:**
Method | GSM | MBPP | HumanEval | ESFT-intent | ESFT-law | ESFT-summary | ESFT-translation
-- | -- | -- | -- | -- | -- | -- | --
Conventional | 53.301 | 32.8 | 37.38 | 81.8 | 29.5 | 39 | 30
DenseMixer | 54.97 | 34 | 39.58 | 83 | 30.75 | 41 | 33.75


When we performed the experiments before the blog release, we directly hacked huggingface transformers library locally. But when we release DenseMixer, to make it easier to use, we reorganized the code and published a pypi package that patches the transformers library.  

There might be some slight difference between these two implementations, which causes the reproduced results to be not exactly the same as the blog shows. But, the conclusion is still the same. :)

**Comment by neilwen987** [comment]
Thanks for share the updated results! This really helps a lot!
